src/editor/codemirror/dnd.ts

```diff
@@ -0,0 +1,77 @@
+import { ChangeSet, Transaction } from "@codemirror/state";
+import { EditorView } from "@codemirror/view";
+import { pythonSnippetMediaType } from "../../common/mediaTypes";
+import { calculateChanges } from "./edits";
+
+// No access to the content via the event in dragover (as it may be cross-document)
+// so we need shared state with the drag so we can draw previews of the drop in
+// the document.
+let draggedCode: string | undefined;
+let invertedChanges: ChangeSet | undefined;
+
+export const setDraggedCode = (value: string | undefined) => {
+  draggedCode = value;
+};
+
+export const snippetDropSupport = () => {
+  let lastPos: number | undefined;
+  const revertPreview = (view: EditorView) => {
+    if (invertedChanges) {
+      view.dispatch({
+        changes: invertedChanges,
+        annotations: [Transaction.addToHistory.of(false)],
+      });
+      invertedChanges = undefined;
+    }
+  };
+  return [
+    EditorView.domEventHandlers({
+      dragover(event, view) {
+        if (draggedCode) {
+          const pos = view.posAtCoords(event);
+          if (pos !== null && pos !== lastPos) {
+            revertPreview(view);
+            lastPos = pos;
+            const transaction = calculateChanges(view.state, draggedCode, pos);
+            // Stash the inverse for when we next update the doc — we'll "undo" first.
+            invertedChanges = transaction.changes.invert(
+              transaction.startState.doc
+            );
+            // Skip selection updates
+            view.dispatch({
+              changes: transaction.changes,
+              annotations: [Transaction.addToHistory.of(false)],
+            });
+          }
+        }
+      },
+      dragleave(event, view) {
+        // dragenter and leave are fired for the child elements of the view too.
+        const rect = view.contentDOM.getBoundingClientRect();
+        if (
+          event.clientY < rect.top ||
+          event.clientY >= rect.bottom ||
+          event.clientX < rect.left ||
+          event.clientX >= rect.right
+        ) {
+          revertPreview(view);
+        }
+      },
+      drop(event, view) {
+        revertPreview(view);
+
+        event.preventDefault();
+        const pos = view.posAtCoords(event);
+        if (pos !== null) {
+          const pythonText = event.dataTransfer?.getData(
+            pythonSnippetMediaType
+          );
+          if (pythonText) {
```

Comment:
The order of this all seems wrong. Media type, then preventDefault, then position?

Reply:
Rearranged as suggested above.